hie-plugin-api/Haskell/Ide/Engine/TypeMap.hs

```diff
@@ -0,0 +1,145 @@
+
+
+-- | Obtain details map for types.
+types :: GHC.HscEnv -> GHC.TypecheckedSource -> IO TypeMap
+types hs_env = everythingInTypecheckedSourceM (ty `combineM` fun)
+ where
+  ty :: forall a . Data a => a -> IO TypeMap
+  ty term = case cast term of
+    (Just lhsExprGhc@(GHC.L (GHC.RealSrcSpan spn) _)) ->
+      getType hs_env lhsExprGhc >>= \case
+        Nothing       -> return IM.empty
+        Just (_, typ) -> return (IM.singleton (rspToInt spn) typ)
+    _ -> return IM.empty
+
+  fun :: forall a . Data a => a -> IO TypeMap
+  fun term = case cast term of
+    (Just (GHC.L (GHC.RealSrcSpan spn) hsPatType)) ->
+      return (IM.singleton (rspToInt spn) (TcHsSyn.hsPatType hsPatType))
+    _ -> return IM.empty
+
+-- | Combine two queries into one using alternative combinator.
+combineM
+  :: (forall a . Data a => a -> IO TypeMap)
+  -> (forall a . Data a => a -> IO TypeMap)
+  -> (forall a . Data a => a -> IO TypeMap)
+combineM f g x = do
+  a <- f x
+  b <- g x
+  return (a `IM.union` b)
+
+-- | Variation of "everything" that does not recurse into children of type t
+-- requires AllowAmbiguousTypes
+everythingButTypeM
+  :: forall t
+   . (Typeable t)
+  => (forall a . Data a => a -> IO TypeMap)
+  -> (forall a . Data a => a -> IO TypeMap)
+everythingButTypeM f = everythingButM $ (,) <$> f <*> isType @t
+
+-- | Returns true if a == t.
+-- requires AllowAmbiguousTypes
+isType :: forall a b . (Typeable a, Typeable b) => b -> Bool
+isType _ = isJust $ eqT @a @b
+
+-- | Variation of "everything" with an added stop condition
+-- Just like 'everything', this is stolen from SYB package.
+everythingButM
+  :: (forall a . Data a => a -> (IO TypeMap, Bool))
+  -> (forall a . Data a => a -> IO TypeMap)
+everythingButM f x = do
+  let (v, stop) = f x
+  if stop
+    then v
+    else Data.gmapQr
+      (\e acc -> do
+        e' <- e
+        a  <- acc
+        return (e' `IM.union` a)
+      )
+      v
+      (everythingButM f)
+      x
+
+-- | Attempts to get the type for expressions in a lazy and cost saving way.
+-- Avoids costly desugaring of Expressions and only obtains the type at the leaf of an expression.
+--
+-- Implementation is taken from: HieAst.hs<https://gitlab.haskell.org/ghc/ghc/blob/1f5cc9dc8aeeafa439d6d12c3c4565ada524b926/compiler/hieFile/HieAst.hs>
+-- Slightly adapted to work for the supported GHC versions 8.2.1 - 8.6.4
+--
+-- See #16233<https://gitlab.haskell.org/ghc/ghc/issues/16233>
+getType
+  :: GHC.HscEnv -> GHC.LHsExpr GhcTc -> IO (Maybe (GHC.SrcSpan, Type.Type))
+getType hs_env e@(GHC.L spn e') =
+  -- Some expression forms have their type immediately available
+  let
+    tyOpt = case e' of
+      HsOverLitType t -> Just t
+      HsLitType t -> Just t
+      HsLamType t -> Just t
+      HsLamCaseType t -> Just t
+      HsCaseType t -> Just t
+      ExplicitListType t -> Just t
+      ExplicitSumType t -> Just t
+      HsMultiIfType t -> Just t
+
+      _                        -> Nothing
+  in  case tyOpt of
```

Comment:
Interesting this isn't pattern matched on.. perhaps this is why you were so confused.

Reply:
It was certainly an additional barrier for me! 
However, since we established that this is a bug, it is actually more clear for me!